README.md
# To better support TPP as well as Venafi as a Service, this project has been renamed to VenafiPS and moved to [a new project](https://github.com/gdbarron/VenafiPS)
This project will stay live for a while, but please submit all issues and questions to the new project.  The [new VenafiPS module](https://www.powershellgallery.com/packages/VenafiPS) has already been published to PowerShellGallery.

----------------------------
----------------------------
----------------------------
# VenafiTppPS - PowerShell module for Venafi Trust Protection Platform and CodeSign Protect

[![Build status](https://gd-barron.visualstudio.com/VenafiTppPS/_apis/build/status/VenafiTppPS)](https://dev.azure.com/gd-barron/VenafiTppPS/_build/latest?definitionId=1)
[![Documentation Status](https://readthedocs.org/projects/venafitppps/badge/?version=latest)](https://venafitppps.readthedocs.io/en/latest/?badge=latest)
[![PowerShell Gallery Version](https://img.shields.io/powershellgallery/v/VenafiTppPS?style=plastic)](https://www.powershellgallery.com/packages/VenafiTppPS)
![PowerShell Gallery](https://img.shields.io/powershellgallery/dt/VenafiTppPS?style=plastic)

## Documentation

Documentation can be found at [http://venafitppps.readthedocs.io](http://venafitppps.readthedocs.io) or by using built-in PowerShell help.  Every effort has been made to document each parameter and provide good examples.

## Supported Platforms

| OS             | PowerShell Version Tested | Status  |
| -------------- |--------------------| -----|
| Windows        | 5.1                | **Working!** |
| Windows        | Core 6.2.3+         | **Working!** |
| MacOS          | Core 6.2.3+         | **Working!** |
| Linux (Ubuntu 18.04) | Core 6.2.3+         | **Working!** |

## Install Module

VenafiTppPS is published to the PowerShell Gallery.  The most recent version is listed in the badge 'powershell gallery' above and can be viewed by clicking on it.  To install the module, you need to have PowerShell installed first.  On Windows, PowerShell will already be installed.  For [Linux](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux?view=powershell-7) or [macOS](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-macos?view=powershell-7), you will need to install PowerShell Core; follow those links for guidance.  Once PowerShell is installed, start a PowerShell prompt and execute `Install-Module -Name VenafiTppPS` which will install from the gallery.

## Usage

Start a new PowerShell prompt (even if you have one from the Install Module step) and create a new VenafiTppPS session with

```powershell
$cred = Get-Credential
New-TppSession -Server 'venafi.mycompany.com' -Credential $cred
```

This will create a session which will be used by default in other functions.
You can also use integrated authentication, simply exclude `-Credential $cred`.  As of v2.0, token-based authentication has been added as well.
View the help on all the ways you can create a new TPP session with `help New-TppSession -full`.

One of the easiest ways to get started is to use `Find-TppObject`:

```powershell
$allPolicy = Find-TppObject -Path '\ved\policy' -Recursive
```

This will return all objects in the Policy folder.  You can also search from the root folder, \ved.

To find a certificate object, not retrieve an actual certificate, use:
```powershell
$cert = Find-TppCertificate -Limit 1
```

Check out the parameters for `Find-TppCertificate` as there's an extensive list to search on.

Now you can take that certificate 'TppObject' and find all log entries associated with it:

```powershell
$cert | Read-TppLog
```

You can also have multiple sessions at once, either to the same server with different credentials or different servers.
This can be helpful to determine the difference between what different users can access or perhaps compare folder structures across environments.  The below will compare the objects one user can see vs. another.

```powershell
# assume you've created 1 session already as shown above...

$user2Cred = Get-Credential # specify credentials for a different/limited user

# get a session as user2 and save the session in a variable
$user2Session = New-TppSession -ServerUrl 'https://venafi.mycompany.com' -Credential $user2Cred -PassThru

# get all objects in the Policy folder for the first user
$all = Find-TppObject -Path '\ved\policy' -Recursive

# get all objects in the Policy folder for user2
$all2 = Find-TppObject -Path '\ved\policy' -Recursive -TppSession $user2Session

Compare-Object -ReferenceObject $all -DifferenceObject $all2 -Property Path
```

## Contributing

Please feel free to log an issue for any new features you would like, bugs you come across, or just simply a question.  I am happy to have people contribute to the codebase as well.


CHANGELOG.md
## 2.2.6
- New project messaging.  Please use VenafiPS going forward.

## 2.2.4
- Add `-KeystorePassword` option to `Get-TppCertificate`.  [#147](https://github.com/gdbarron/VenafiTppPS/issues/147).  Thanks @Curtmcgirt!

## 2.2.3
- Fix [#145](https://github.com/gdbarron/VenafiTppPS/issues/145), `Revoke-TppToken` doesn't show target.  Thanks @wilddev65!

## 2.2.2
- Rename 'Provision' to 'Push', aliases added for existing code
- Add `Invoke-TppCertificatePush`
- Fix [#130](https://github.com/gdbarron/VenafiTppPS/issues/130), `Get-TppDevice` only accepting IP address for host, not hostname.  Thanks @Curtmcgirt!
- Fix [#131](https://github.com/gdbarron/VenafiTppPS/issues/131), add examples to `New-TppCapiApplication`.  Thanks @Curtmcgirt!
- Fix [#132](https://github.com/gdbarron/VenafiTppPS/issues/132), 500 error setting BindingIpAddress running `New-TppCapiApplication`.    Thanks @Curtmcgirt!
- Fix [#134](https://github.com/gdbarron/VenafiTppPS/issues/134), server url is blank when running `Get-TppObject` with secondary token.  This was an issue for `Get-TppPermission` as well.  Thanks @stevekeever!
- Add missing parameters comment-based help for `New-TppCapiApplication`
- Fix certificate push not working in `New-TppCapiApplication`
- Update links to reference `main` branch instead of `master`

## v2.2.0
- Identity format validation fix, [#126](https://github.com/gdbarron/VenafiTppPS/issues/126).  Thanks @DadsVacayShorts!
- Add `Get-TppIdentity` to retrieve Identity info given an id
- Add `Remove-TppPermission`, accepts output from `Get-TppPermission`
- Add Path param to `Set-TppPermission` in addition to guid
- `Get-TppPermission` now accepts TppObject, eg. from `Find-TppObject`
- `Set-TppPermission` now accepts output from `Get-TppPermission` for the object and IdentityId so you only need to specify Permission. No need to get guid and identity manually to pass in.
- `Find-TppIdentity` output standardized so you can now pipe to permission functions
- `Get-TppPermission` returns additional object and identity info
- Centralize format validation for identities
- Update help links referring to versions no longer available
- `Find-TppIdentity -Me` to be deprecated for `Get-TppIdentity -Me`
- Add option to `Get-TppObject` for guid
- Standardized on Id/IdentityId for the identity across all identity and permission functions
- Force missing slash retry to status codes of only 307 and 401
- Better error handling and messaging through the permission functions

## v2.1.1
- `Get-TppPermission` fix when retrieving multiple permissions, [#124](https://github.com/gdbarron/VenafiTppPS/issues/124).  Thanks @DadsVacayShorts!

## v2.1.0
- Update `Get-TppCertificateDetail` help to ensure output lists the correct properties, [#119](https://github.com/gdbarron/VenafiTppPS/issues/119).  Thanks @doyle043!
- Hide secret info, eg. passwords, tokens, etc, when verbose logging.  [#120](https://github.com/gdbarron/VenafiTppPS/issues/120).  Thanks @bwright86!
- Add search, get, and remove code sign project and environment functions
- Fix, provide the correct error message when making rest call and testing to see if a trailing slash is needed or not
- Update `New-TppSession` to ensure $TppSession is created even if subsequent custom field calls fail
- Update TppSession object Validate method to check if token auth is required.  Needed for code sign.

## v2.0.5
- Add missing filters CreateDate, CreatedBefore, and CreatedAfter to `Find-TppCertificate`, [#117](https://github.com/gdbarron/VenafiTppPS/issues/117).  Thanks @doyle043!

## v2.0.4
- Fix header getting stripped causing `Write-TppLog` to fail, [#114](https://github.com/gdbarron/VenafiTppPS/issues/114).  Thanks @stevekeever!
- Update `Invoke-TppRestMethod` to retry with trailing slash for all methods, not just Get

## v2.0.3
- Add Origin property when creating a new certificate
- Add icon to project, [#37](https://github.com/gdbarron/VenafiTppPS/issues/37)

## v2.0.2
- Process to convert a secure password to plain text was failing on Linux, [#108](https://github.com/gdbarron/VenafiTppPS/issues/108).  Thanks @macflurry7!

## v2.0.1
- Add Import-TppCertificate, [#88](https://github.com/gdbarron/VenafiTppPS/issues/88).  Thanks @smokey7722!
- Make Invoke-TppRestMethod accessible, [#106](https://github.com/gdbarron/VenafiTppPS/issues/106).  Thanks @wilddev65!
- Fix verbose being turned on incorrectly in New-TppSession when getting by token

## v2.0.0
- Add token-based authentication support, Integrated, OAuth, and Certificate. Tokens can be used in or out of this module. [#94](https://github.com/gdbarron/VenafiTppPS/issues/94).  Thanks @BeardedPrincess!
- Add CertificateType option to New-TppCertificate
- Add support for GET api calls which require a trailing slash
- Fixes in multiple functions where .Add on a hashtable was called in the process block
- Fix issue #102, Base64 with private key not an available option
- Update formats which support IncludeChain

## v1.2.5
- Add offset parameter to Find-TppCertificate, [#92](https://github.com/gdbarron/VenafiTppPS/issues/92)
- Allow inclusion of private key for format Base64 (PKCS #8) in Get-TppCertificate.  Earlier versions of Venafi documentation listed this incorrectly, but has been resolved.  [#95](https://github.com/gdbarron/VenafiTppPS/issues/95)
- Get-TppCertificate failing when pipilining due to adding a key to a hashtable that already exists, [#96](https://github.com/gdbarron/VenafiTppPS/issues/96)
- Linux style paths which use / instead of \ were failing path check due to invalid regex, [#97](https://github.com/gdbarron/VenafiTppPS/issues/97)
- PSSA fix for Read-TppLog

## v1.2.3
- ProvisionCertificate not triggering a push, [#89](https://github.com/gdbarron/VenafiTppPS/issues/89)

## v1.2.2
- Add Linux support
- Add New-TppDevice
- New-TppCapiApplication
    - Add ProvisionCertificate parameter to provision a certificate when the application is created

VenafiTppPS/Enums/ConfigResult.ps1
enum ConfigResult {
    Success = 1
    InvalidArgument = 2
    InvalidArgumentRange = 3
    MismatchedArguments = 4
    NotImplemented = 5
    InvalidDestinationList = 6
    InsufficientPrivileges = 7
    InvalidOperation = 8
    UnexpectedAssemblyError = 9
    OutOfMemory = 10
    AttributeDoesNotExist = 100
    AttributeAlreadyExists = 101
    AttributeNotFound = 102
    AttributeValueExists = 103
    AttributeStillInUse = 104
    AttributeNameTooLong = 105
    AttributeReferenceDoesNotExist = 106
    AttributeSyntaxCollision = 107
    AttributePropertyCollision = 108
    CannotRemoveMandatory = 109
    AttributeValueIsMandatory = 110
    AttributeValueTooLong = 111
    IllegalAttributeForClass = 112
    InvalidAttributeDN = 113
    AttributeValueDoesNotExist = 114
    AttributeIsSingleValued = 115
    AttributeIsReadOnly = 116
    AttributeIsHidden = 117
    ClassDoesNotExist = 200
    ClassAlreadyExists = 201
    ClassStillInUse = 202
    ClassNameTooLong = 203
    ClassInvalidSuperClass = 204
    ClassInvalidContainmentClass = 205
    ClassInvalidNamingAttribute = 206
    ClassInvalidMandatoryAttribute = 207
    ClassInvalidOptionalAttribute = 208
    ClassInvalidName = 209
    ClassInvalidContainmentSubClass = 210
    PolicyDoesNotExist = 300
    PolicyLockStateCollision = 301
    LockNameAlreadyExists = 350
    LockNameDoesNotExist = 351
    LockNameOwnedByAnother = 352
    LockNameLimitReached = 353
    LockNameAttemptTimedOut = 354
    ObjectDoesNotExist = 400
    ObjectAlreadyExists = 401
    ObjectHasChildren = 402
    ObjectNameTooLong = 403
    ObjectDepthTooDeep = 404
    ObjectInvalidName = 405
    ObjectInvalidClass = 406
    ObjectInvalidContainment = 407
    ObjectMandatoryMissing = 408
    ObjectIsReadOnly = 409
    ObjectInvalidOperation = 410
    DriverMissingDSN = 500
    DriverMissingDatabaseName = 501
    DriverDatabaseError = 502
    DriverTransactionError = 503
    DriverTransactionCollision = 504
    DriverGenerationUpdateError = 505
    CacheLockException = 600
    CacheEntryNotFound = 601
    CacheEntryAlreadyExists = 602
    CacheEntryIsSuperior = 603
    CacheEntryIsIncompatible = 604
    XmlInvalidStructure = 700
    XmlMissingNaming = 701
    XmlMissingSyntax = 702
    XmlMissingProperty = 703
    XmlUnknownElementAttribute = 704
}
